Fix #823: Question content item Hi-fi

nikitamarysolomanpvt opened the pull request:
## Explanation
Question Player Content item UI update -Hi-fi.
Duplicate of #824, changes in alignment with respect to recycler-view.
## Mock
https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/8fbb663a-39aa-4be1-83bd-73bde6f85dcc/PM-Q1-Fraction-Input-No-Answer-
## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

vinitamurthi committed:
Initial check in for the question data controller

vinitamurthi committed:
Removed progress controller test

BenHenning committed:
Address the renames suggested in #217.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-interface

BenHenning committed:
Introduce thoroughly stubbed QuestionAssessmentProgressController +
empty test suite.

vinitamurthi committed:
Review changes

vinitamurthi committed:
Review changes

vinitamurthi committed:
Pass data providers to the assessment controller, and set caps in the training controller

vinitamurthi committed:
Fixed merge conflicts

vinitamurthi committed:
Removed unnecessary imports/variables

vinitamurthi committed:
Add fake question data for stubbed interfaces

vinitamurthi committed:
Remove duplicate questions while fetching in training controller

vinitamurthi committed:
Comment explaining the filter function

vinitamurthi committed:
Improve duplicate checking - check it while filtering instead of after filtering

vinitamurthi committed:
Add linked skill id values

vinitamurthi committed:
Review changes

vinitamurthi committed:
add a new module for questions constants

vinitamurthi committed:
fix merge conflicts

vinitamurthi committed:
Review changes

vinitamurthi committed:
add a test to verify questions were fetched properly

vinitamurthi committed:
reformatted code

BenHenning committed:
Merge branch 'question_data' into introduce-question-progress-controller-interface

Conflicts:
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt
	model/src/main/proto/question.proto

BenHenning committed:
Re-add QuestionTrainingController removed in merge.

BenHenning committed:
Finalize question progress controller interface.

BenHenning committed:
Implement QuestionAssessmentProgressController.

This includes some basic refactoring of internal structures used by the
exploration progress controller to share common functionality between the
two progress controllers. There's still some duplication, but this seems
like a reasonable split since there's likely to be further differences in
the progress controllers in the future.

The question assessment progress controller tests pass, but no new ones
have yet been added to thoroughly test the implementation.

BenHenning committed:
Add initial phase of tests for the progress controller based on
ExplorationProgressController. They haven't yet been verified as correct.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-interface

Conflicts:
	domain/src/main/assets/sample_questions.json
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/main/java/org/oppia/domain/util/JsonAssetRetriever.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Fix typo in QuestionTrainingController.

BenHenning committed:
Post-merge fixes and adjustments.

BenHenning committed:
Fix broken ProfileManagementControllerTest.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-interface' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Merge branch 'fix-broken-profile-management-controller-test' into introduce-question-progress-controller-implementation

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Update tests to build, but not pass.

jamesxu0 committed:
Merge branch 'introduce-question-progress-controller-interface' into question_player_lofi

jamesxu0 committed:
Started UI

jamesxu0 committed:
setup ViewModel

jamesxu0 committed:
added binding adapter

jamesxu0 committed:
Handles basic functionality

jamesxu0 committed:
Moved code around

jamesxu0 committed:
Added replay button

jamesxu0 committed:
Merge branch 'develop' into question_player_lofi

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/main/java/org/oppia/domain/util/JsonAssetRetriever.kt
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Merge branch 'develop' into question_player_lofi

Conflicts:
	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
	app/src/main/res/layout/question_player_fragment.xml
	app/src/sharedTest/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivityTest.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into question_player_lofi

Conflicts:
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt

BenHenning committed:
Make questions partially work.

This introduces back button behavior to ensure training sessions can be
reset.

This also attempts to introduce proper asset loading, but it doesn't
seem to be working currently.

BenHenning committed:
Temporarily remove check in StateDeck that breaks questions.

BenHenning committed:
Ensure all state navigation buttons are on the same line and the same
size.

This required reworking how nav button binding works in the state
fragment such that multiple models are used depending on which type of
navigation scenario the fragment is currently in.

BenHenning committed:
Fix dp adjustment that caused previous button to disappear.

Fix previous button showing up twice for continue interaction.

BenHenning committed:
Undo .idea/misc.xml change that snuck in.

BenHenning committed:
Update & fix StateFragmentTest (at least for Espresso). See #495 description
for details on the changes necessary to make these tests work.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into question_player_lofi

Conflicts:
	app/src/main/res/values/colors.xml

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into temp-base-branch-for-question-player

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

BenHenning committed:
Post-merge fixes with some temporary stop-gap solutions.

BenHenning committed:
Fix questions image rendering and consolidate GCS resource bucket names.

BenHenning committed:
Fix question session not properly being reinitialized for new sessions.

Also, ensure that notifications do not regenerate the training session.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

BenHenning committed:
Fix index accounting for questions. Before, the index incremented upon
correct answer submission rather than navigation.

BenHenning committed:
Remove support for backward navigation in questions since it isn't
allowed.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

Conflicts:
	domain/src/main/java/org/oppia/domain/state/StateDeck.kt

BenHenning committed:
Set title of question activity to 'Train Mode'.

BenHenning committed:
Add support for a synthetic ephemeral question to represent the end of a
training session.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

Conflicts:
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	model/src/main/proto/question.proto

BenHenning committed:
Update question progress tracking to be translatable and to ensure that
the synthetic terminal state shows the correct progress: 'completed'.

BenHenning committed:
Move most state player functional support into a new configurable
assembler class to enable sharing between the exploration and question
players.

BenHenning committed:
Hook up StatePlayerRecyclerViewAssembler to the question player fragment
presenter. This also enables the congratulations text.

BenHenning committed:
Introduce proper terminal page and disable overscroll. Note that the end
session page introduced here has not been mocked yet, so it's mostly a
placeholder until a mock is available.

BenHenning committed:
Add support & button for training session replay.

BenHenning committed:
Update/add TODOs to correspond to GitHub tracked issues.

BenHenning committed:
Fix broken import.

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

BenHenning committed:
Finish tests for Question{Training,AssessmentProgress}Controllers.

Note that the nature of the assessment progress controller is such that
a new, dynamic DataProvider is needed since the data provider for the
current assessment question is actually a transformation of different
data providers depending on which questions are passed to the
controller.

BenHenning committed:
Initial introduction of test coroutine dispatchers to replace Kotlin's
test coroutine dispatcher.

This includes introducing a test-only module to contain testing
dependencies.

BenHenning committed:
Introduce a new LiveData mechanism to bridge coroutines from
DataProviders in a way that doesn't have the same limitations as the
previous MutableLiveData-based bridge.

BenHenning committed:
Merge branch 'develop' into introduce-test-coroutine-dispatchers

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

Conflicts:
	domain/src/test/java/org/oppia/domain/audio/CellularAudioDialogControllerTest.kt

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-question-progress-controller-implementation

BenHenning committed:
Introduce new nested data provider that effectively allows a
DataProvider to be set up like transformAsync() but with the ability to
change the root DataProvider.

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

nikitamarysolomanpvt committed:
question player toolbar shadow

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
question content for question player

nikitamarysolomanpvt committed:
question player progress-bar hifi

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
revert

nikitamarysolomanpvt committed:
revert

nikitamarysolomanpvt committed:
Merge remote-tracking branch 'upstream/introduce-question-player-ui-structure' into question-player-progress-bar-hifi

nikitamarysolomanpvt committed:
spcacing fix with respect to recycler view spacing

rt4914 reviewed (approved):
LGTM thanks.

BenHenning committed:
Early work at introducing FakeSystemClock tests (not yet complete).

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

BenHenning committed:
Merge branch 'develop' into introduce-test-coroutine-dispatchers

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

Conflicts:
	utility/src/main/java/org/oppia/util/data/DataProviders.kt

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

BenHenning committed:
Merge branch 'develop' into align-and-dedup-continue-button-state-fragment

Conflicts:
	app/src/main/AndroidManifest.xml
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt
	app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragment.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/answerhandling/InteractionAnswerHandler.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelFactory.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelModule.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt
	app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt
	app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivityPresenter.kt
	app/src/main/java/org/oppia/app/testing/ContentCardTestActivity.kt
	app/src/main/java/org/oppia/app/testing/ContentCardTestActivityPresenter.kt
	app/src/main/res/layout/my_downloads_activity.xml
	app/src/main/res/layout/state_button_item.xml
	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/app/topic/overview/TopicOverviewFragmentTest.kt

nikitamarysolomanpvt committed:
nit

rt4914 commented on app/src/main/res/layout/question_player_fragment.xml:
One blank line above this.

Also, to your recyclerview add `clipToPadding as false` otherwise your UI will get cut while scrolling:
<img width="473" alt="Screenshot 2020-04-03 at 12 13 40 PM" src="https://user-images.githubusercontent.com/9396084/78331878-8eda9f00-75a4-11ea-8fd6-dbd084d3a22a.png">

It was done in landscape mode just to show this scenario but it can happen in portrait mode also if the content is large enough.

rt4914 commented on app/src/main/res/layout/question_player_content_item.xml:
The top margin in mocks is 38dp which can be (36 or 40). Now your recyclerview contains a padding of `16dp` and therefore this top margin should be either `20dp` or `24dp`

rt4914 commented on app/src/main/res/layout/question_player_content_item.xml:
This should not be there as you have already covered margin in frame layout.

rt4914 commented on app/src/main/res/layout/question_player_content_item.xml:
This should also be removed otherwise the next item will be at `44dp` margin, 16 from textview + 28 from framelayout

rt4914 reviewed (changes requested):
Changes suggested. PTAL

nikitamarysolomanpvt commented on app/src/main/res/layout/question_player_content_item.xml:
removed recycler view padding as per other interactions spacing updated margin.. hence  
 ```
android:layout_marginTop="36dp"
android:layout_marginBottom="28dp"
```
bottom margin will be added with margin 8dp of upcoming items in recyclerview

nikitamarysolomanpvt commented on app/src/main/res/layout/question_player_content_item.xml:
ditto

nikitamarysolomanpvt commented on app/src/main/res/layout/question_player_content_item.xml:
ditto

nikitamarysolomanpvt committed:
nit removed recycler view padding as per other interactions spacing updated margin.. hence

    android:layout_marginTop="36dp"
    android:layout_marginBottom="28dp"
    bottom margin will be added with margin 8dp of upcoming items in recyclerview

rt4914 reviewed (approved):
LGTM, thanks. now the padding/margins are correct.

nikitamarysolomanpvt commented on app/src/main/res/layout/question_player_fragment.xml:
padding is removed

nikitamarysolomanpvt committed:
nit

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into introduce-question-player-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragment.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateViewModel.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelFactory.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelModule.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt
	app/src/main/java/org/oppia/app/player/stopplaying/StopExplorationDialogFragment.kt
	app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt
	app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerFragment.kt
	app/src/main/res/values/colors.xml
	app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt
	domain/src/main/java/org/oppia/domain/topic/TopicListController.kt
	domain/src/test/java/org/oppia/domain/topic/TopicListControllerTest.kt
	utility/src/main/java/org/oppia/util/parser/HtmlParser.kt

BenHenning committed:
Merge branch 'develop' into align-and-dedup-continue-button-state-fragment

Conflicts:
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/res/layout/state_button_item.xml
	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Fix the submit button disabled states: with the new refactor, the submit
button is now properly disabled when an invalid answer is pending.

BenHenning committed:
Revert accidental changes to .idea/misc.xml.

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Merge branch 'develop' into introduce-test-coroutine-dispatchers

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

Conflicts:
	data/src/test/java/org/oppia/data/persistence/PersistentCacheStoreTest.kt

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Remove unnecessary meta file for Mockito.

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into temp-base-branch-for-question-player

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Minor import cleanup.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into introduce-question-player-ui-structure

This merge resolution includes a simplification/re-write of hint &
solution handling for explorations (and in a way that might be
easily transferable to questions).

Conflicts:
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateViewModel.kt
	app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Fix audio playback & generalize it to facilitate enabling it for
questions in the future.

BenHenning committed:
Address reviewer comments & fix broken tests.

BenHenning committed:
Fix hints & solutions functionality after simplification (some bugs &
regressions were introduced due to the simplification; these have been
fixed).

BenHenning committed:
Revert the assembler delays to match the intended durations rather than
the test-only values used for development.

BenHenning committed:
Merge branch 'develop' into align-and-dedup-continue-button-state-fragment

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Remove infeasible testing structures, add documentation, and clean up
implementation to prepare for code review.

BenHenning committed:
Add notice that the dispatchers utility is temporary.

BenHenning committed:
Cleanup new LiveData bridge, add tests for it, and migrate other
DataProviders tests to using TestCoroutineDispatchers utility.

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Add AsyncResult tests, fix & re-enable an earlier test in PersistentCacheStoreTest, and fix FakeSystemClock so that it works properly in test environments.

BenHenning committed:
Use ktlint to reformat TestCoroutineDispatchers per reviewer comment
thread.

BenHenning committed:
Merge branch 'introduce-test-coroutine-dispatchers' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Reformat files failing linter check.

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

Conflicts:
	utility/src/main/java/org/oppia/util/data/DataProviders.kt

BenHenning committed:
Reformat new DataProviders code.

BenHenning committed:
Add new tests for the NestedTransformedDataProvider.

BenHenning committed:
Merge branch 'develop' into introduce-new-coroutine-live-data-mechanism

 Conflicts:
	testing/src/main/java/org/oppia/testing/FakeSystemClock.kt

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Merge branch 'align-and-dedup-continue-button-state-fragment' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

Conflicts:
	domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into introduce-question-player-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/hintsandsolution/HintsAndSolutionAdapter.kt
	utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt

BenHenning committed:
Address reviewer comments & fix lint issue.

BenHenning committed:
Merge branch 'develop' into introduce-new-coroutine-live-data-mechanism

BenHenning committed:
Merge branch 'introduce-new-coroutine-live-data-mechanism' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'develop' into introduce-nested-data-provider

BenHenning committed:
Merge branch 'introduce-nested-data-provider' into introduce-question-progress-controller-implementation

BenHenning committed:
Merge branch 'develop' into introduce-question-progress-controller-implementation

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into introduce-question-player-ui-structure

BenHenning committed:
Update hints & solutions handling to more closely follow the intended
behavior graph (and added an actual visual representation of this graph
in code to simplify maintaining the solution in the future). Tests will
be added as part of solving #1273.

BenHenning committed:
Merge branch 'develop' into introduce-question-player-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt

BenHenning committed:
Merge branch 'introduce-question-player-ui-structure' into question-player-question-content-hifi

Conflicts:
	app/src/main/java/org/oppia/app/player/state/StatePlayerRecyclerViewAssembler.kt

BenHenning committed:
Merge branch 'question-player-question-content-hifi' of github.com:oppia/oppia-android into question-player-question-content-hifi

BenHenning committed:
Merge branch 'develop' into introduce-question-player-ui-structure

BenHenning committed:
Merge branch 'introduce-question-player-ui-structure' into question-player-question-content-hifi

veena14cs commented on app/src/main/res/layout/question_player_content_item.xml:
I think margin bottom is 36dp in the mock.

veena14cs reviewed (approved):
Please address one comment and it is good to merge.

nikitamarysolomanpvt commented on app/src/main/res/layout/question_player_content_item.xml:
This will be added to divider_margin_top i.e. 8 dp in the interactions and response.

veena14cs commented on app/src/main/res/layout/question_player_content_item.xml:
Got it. thanks.

nikitamarysolomanpvt commented:
Duplicate of #1325
